sssd_test_framework/roles/ad.py

```diff
@@ -1490,4 +1545,216 @@ def __get_info(
+
+        :param target: Group policy target, defaults to 'Default-First-Site-Name'.
+        :type target: str, optional
+        :return: Group policy object
+        :rtype: GPO
+
+        Group policies objects needs to be be linked to OUs, domains or sites.
+        """
+        self.role.host.ssh.run(f'New-GPLink -Guid "{self._cn}" -Target "{target}" -LinkEnabled Yes')
+        self.host._backup_gpo.append(f'Remove-GPLink -Guid "{self._cn}" -Target "{target}"')
+
+        return self
+
+    def unlink(self, target: str | None = "Default-First-Site-Name") -> GPO:
+        """
+        Unlink the group policy from the target.
+
+        :param target: Group policy target, defaults to 'Default-First-Site-Name'.
+        :type target: str, optional
+        :return: Group policy object
+        :rtype: GPO
+        """
+        self.role.host.ssh.run(f'Remove-GPLink -Guid "{self._cn}" -Target "{target}"')
+        self.host._backup_gpo.remove(f'Remove-GPLink -Guid "{self._cn}" -Target "{target}"')
+
+        return self
+
+    def permissions(self, level: str, target_type: str | None = "User") -> GPO:
+        """Configure group policy object permissions.
+
+        :param level: Permission level
+        :type level: str, values can be 'GpoRead | GpoApply | GpoEdit | GpoEditDeleteModifySecurity | None'
+        :param target_type: Target type, defaults to 'user'
+        :type target_type: str, optional
+        :return: Group policy object
+        :rtype: GPO
+        """
+        self.role.host.ssh.run(
+            f'Set-GPPermission -Guid "{self._cn}" -PermissionLevel {level} -Type "{target_type}" -Replace $True'
+        )
+
+        return self
+
+    def policy(self, logon_rights: dict[str, list[ADObject]]) -> GPO:
+        """
+        Group policy configuration.
+
+        This method does the remaining configuration of the group policy. It updates
+        `GptTmpl.inf` with security logon right keys with the SIDs of users and groups
+        objects.
+
+        To add users and groups to the policy, the SID must be used for the values. The
+        value constructed will look like; '*SID-1-2-3,*SID-1-2-3-4'
+
+        Additionally, gPCMachineExtensionNames need to be updated in the directory so
+        the GPO is readable to the client. The value is a list of Client Side
+        Extensions (CSEs), that is an index of what part of the policy is pushed and
+        processed by the client.
+
+        :param logon_rights: List of logon rights.
+        :type logon_rights: dict[str, list[ADObjects]]
+        :return: Group policy object
+        :rtype: GPO
+        """
+        _keys: list[str] = [
+            "SeInteractiveLogonRight",
+            "SeRemoteInteractiveLogonRight",
+            "SeDenyInteractiveLogonRight",
+            "SeDenyRemoteInteractiveLogonRight",
+        ]
+
+        for i in _keys:
+            if i not in logon_rights.keys():
+                raise KeyError
```

Comment:
Please, provide message to the KeyError. E.g. `f"Expected {_keys} but got {i}`.

You can also use the Literal type instead of str, e.g. `Literal["SeInteractiveLogonRight", ...]` this way it will be checked by mypy.

Reply:
I spent a couple of hours learning and trying to implement Literal, then started poking at TypedDict. I updated it to provide a message for now.